pkg/spanconfig/spanconfigsqlwatcher/sqlwatcher.go

```diff
@@ -0,0 +1,300 @@
+			select {
+			case <-ctx.Done():
+			case errCh <- err:
+			}
+		}
+	}
+
+	err := s.watchForDescriptorUpdates(ctx, timestamp, onEvent, onFrontierAdvance)
+	if err != nil {
+		return errors.Wrapf(err, "error establishing rangefeed over system.descriptors")
+	}
+	err = s.watchForZoneConfigUpdates(ctx, timestamp, onEvent, onFrontierAdvance)
+	if err != nil {
+		return errors.Wrapf(err, "error establishing rangefeed over system.zones")
+	}
+	// We always tear down rangefeeds before returning to the caller.
+	defer s.close()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case err = <-errCh:
+			return err
+		case <-frontierAdvanced:
+			events, combinedFrontierTS := s.buffer.flush(ctx)
+			if len(events) != 0 {
+				if err := handle(ctx, events, combinedFrontierTS); err != nil {
+					return err
+				}
+			}
+		}
+	}
+}
+
+// close stops the rangefeeds created by the SQLWatcher and waits for them
+// to shut down before returning. Close is idempotent.
+func (s *SQLWatcher) close() {
+	s.zonesRF.Close()
+	s.descriptorsRF.Close()
+}
+
+// watchForDescriptorUpdates establishes a rangefeed over system.descriptors and
+// invokes the onEvent callback for observed events. The onFrontierAdvance
+// callback is invoked whenever the rangefeed frontier is advanced as well.
+func (s *SQLWatcher) watchForDescriptorUpdates(
+	ctx context.Context,
+	timestamp hlc.Timestamp,
+	onEvent func(context.Context, event),
+	onFrontierAdvance func(context.Context, rangefeedKind, hlc.Timestamp),
+) error {
+	descriptorTableStart := s.codec.TablePrefix(keys.DescriptorTableID)
+	descriptorTableSpan := roachpb.Span{
+		Key:    descriptorTableStart,
+		EndKey: descriptorTableStart.PrefixEnd(),
+	}
+	handleEvent := func(ctx context.Context, ev *roachpb.RangeFeedValue) {
+		if !ev.Value.IsPresent() && !ev.PrevValue.IsPresent() {
+			// Event for a tombstone on a tombstone -- nothing for us to do here.
+			return
+		}
+		value := ev.Value
+		if !ev.Value.IsPresent() {
+			// The descriptor was deleted.
+			value = ev.PrevValue
+		}
+
+		var descriptor descpb.Descriptor
+		if err := value.GetProto(&descriptor); err != nil {
+			logcrash.ReportOrPanic(
+				ctx,
+				&s.settings.SV,
+				"%s: failed to unmarshal descriptor %v",
+				ev.Key,
+				value,
+			)
+			return
+		}
+		if descriptor.Union == nil {
+			return
+		}
+
+		table, database, typ, schema := descpb.FromDescriptorWithMVCCTimestamp(&descriptor, value.Timestamp)
+
+		var id descpb.ID
+		var descType catalog.DescriptorType
+		switch {
+		case table != nil:
+			id = table.GetID()
+			descType = catalog.Table
+		case database != nil:
+			id = database.GetID()
+			descType = catalog.Database
+		case typ != nil:
+			id = typ.GetID()
+			descType = catalog.Type
+		case schema != nil:
+			id = schema.GetID()
+			descType = catalog.Schema
+		default:
+			logcrash.ReportOrPanic(ctx, &s.settings.SV, "unknown descriptor unmarshalled %v", descriptor)
+		}
+
+		rangefeedEvent := event{
+			timestamp: timestamp,
```

Comment:
Tiny bug I was running into in #71994:

```suggestion
			timestamp: ev.Value.Timestamp,
```

If you can think up the kind of testing that would've caught this, I'll encourage you to add it.